Clarify goal of the spec + trailing commas seperate page (#7)
DecimalTurn/JSONC: trailingcommas.markdown

---

layout: default
---

# Trailing Commas and JSONC

## Why Trailing Commas Aren't Part of the JSONC Specification?

Trailing commas are not part of the JSONC Specification because the reference implementation, [jsonc-parser](https://www.npmjs.com/package/jsonc-parser), does not allow them unless explicitly configured. The `allowTrailingComma` option is set to `false` by default, so any trailing comma will result in a parsing error. 

> The reason why this specification chose the default behavior of the parser as the reference for the standard is to ensure that JSONC remains compatible with the broader JSON ecosystem, which does not allow trailing commas. This decision helps maintain consistency and predictability across different parsers and implementations. Namely, the tsconfig and eslint config files, which are widely used in the JavaScript ecosystem, do not allow trailing commas in their JSONC files.

Reviewer: 1. I believe you meant _"consistent"_ instead of _"compatible"_, or _broader JSONC eocsystem_ instead of _broader _JSON_ ecosystem_. We both know that JSONC's ecosystem is, by its very nature of the format itself, fundamentally incompatible with JSON's ecosytem.
2. Formatting nits:
```suggestion
The reason why this specification chose the default behavior of the parser as the reference for the standard is to ensure that JSONC remains compatible with the broader JSON ecosystem, which does not allow trailing commas. This decision helps maintain consistency and predictability across different parsers and implementations. Namely, [TSConfig](https://www.typescriptlang.org/tsconfig/) and [ESLint config](https://eslint.org/docs/latest/use/configure/configuration-files) files, widespread in the JavaScript ecosystem, do not support trailing commas when stored as JSONC.
```

Reviewer 2: I find the use of "support" and "stored" less clear here, "widespread" I find less clear as well. I prefer the original wording, it seems more on point. The formatting changes are nice, though.

Author: 1. In this case, I really meant compatible in the sense that a config file that can't be read by a certain tool is incompatible with it. Here I'm using "Javascript Ecosystem" as an umbrella term to include JavaScript, TypeScript, JSON and JSONC since all those are closely intertwined.
2. I appreciate the suggestion to add the links to the TSCondig and ESLint pages and I'll include them in a seperate commit. but keep the original wording in that case for simplicity.

Reviewer 2: Hm, but JSONC will never be compatible with JSON, like @Alhadis pointed out, since it's a superset. I agree that point seems confusing and not obviously relevant for trailing commas.

The exclusion of trailing commas also facilitates the creation of tools and libraries that can parse JSONC without needing to handle additional syntax variations. This helps ensure that JSONC remains a lightweight and straightforward extension of JSON, primarily focused on adding comments without introducing significant complexity.

## Can a Parser Choose to Support Trailing Commas?

> Yes, a parser can choose to support trailing commas in JSONC files for convenience. This is especially useful in configuration files or code where lists and objects are frequently edited. Allowing trailing commas can make adding, removing, or reordering items easier and reduce the likelihood of syntax errors. 

Reviewer: This paragraph is sending mixed messages: a parser that chooses to support trailing commas "for convenience" encourages users to use them in their JSONC files, which is at odds with the previous paragraph's message.

Reviewer 2: Good catch, I actually didn't notice this paragraph being added. I agree that it makes more sense not to add it, no matter if trailing commas end up in the spec as allowed or not. (We all know my thoughts on that one.)

Author: I've removed the paragraph and clarified the question in 9ffa0a4f4eea7c62ae4fd9c7f87984dad84a9f96.
Hopefully this makes things clearer in terms of the message here.

Reviewer 2: I like the new clarity, even though I hope the VS Code parser team wakes up soon so we can revisit this choice.
